# **Segmenting and Clustering Neighborhoods in Toronto**
by Yong Chang *(ychang2000@gmail.com)*

### **Part I** -- Build Dataframe

First, let's import the data from wiki webpage and do a quick check to verify the data acquisition.

In [1]:
import pandas as pd
import html5lib
url='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
tables = pd.read_html(url)
table = tables[0]

In [2]:
table.head()

Postcode           Borough     Neighbourhood
0      M1A      Not assigned      Not assigned
1      M2A      Not assigned      Not assigned
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront

In [3]:
table.tail()

Postcode       Borough          Neighbourhood
283      M8Z     Etobicoke              Mimico NW
284      M8Z     Etobicoke     The Queensway West
285      M8Z     Etobicoke  Royal York South West
286      M8Z     Etobicoke         South of Bloor
287      M9Z  Not assigned           Not assigned

In [4]:
table.shape

(288, 3)

In [5]:
table.describe(include='all')

Postcode       Borough Neighbourhood
count       288           288           288
unique      180            12           209
top         M9V  Not assigned  Not assigned
freq          8            77            78



Next, let's work on the data wrangling. As being asked, we will ***remove*** the cells with a ***borough*** that is ***'Not assigned'***.

In [6]:
table_temp1=table.loc[table['Borough']!='Not assigned'].reset_index(drop=True)
table_temp1

Postcode           Borough             Neighbourhood
0        M3A        North York                 Parkwoods
1        M4A        North York          Victoria Village
2        M5A  Downtown Toronto              Harbourfront
3        M5A  Downtown Toronto               Regent Park
4        M6A        North York          Lawrence Heights
..       ...               ...                       ...
206      M8Z         Etobicoke  Kingsway Park South West
207      M8Z         Etobicoke                 Mimico NW
208      M8Z         Etobicoke        The Queensway West
209      M8Z         Etobicoke     Royal York South West
210      M8Z         Etobicoke            South of Bloor

[211 rows x 3 columns]

Then, for neighborhoods having the same postal code, **combine** them separated with a comma.

In [7]:
table_temp2=table_temp1.groupby('Postcode',as_index=False).agg(lambda x:','.join(set(x)))
table_temp2.tail(20)

Postcode       Borough                                      Neighbourhood
83       M6R  West Toronto                              Parkdale,Roncesvalles
84       M6S  West Toronto                                  Swansea,Runnymede
85       M7A  Queen's Park                                       Not assigned
86       M7R   Mississauga              Canada Post Gateway Processing Centre
87       M7Y  East Toronto  Business Reply Mail Processing Centre 969 Eastern
88       M8V     Etobicoke         Mimico South,New Toronto,Humber Bay Shores
89       M8W     Etobicoke                              Alderwood,Long Branch
90       M8X     Etobicoke        The Kingsway,Montgomery Road,Old Mill North
91       M8Y     Etobicoke  Sunnylea,Kingsway Park South East,Mimico NE,Th...
92       M8Z     Etobicoke  Kingsway Park South West,The Queensway West,Mi...
93       M9A     Etobicoke                                   Islington Avenue
94       M9B     Etobicoke  Princess Gardens,West Deane Park,Martin Grove,...
95       M9C     Etobicoke  Markland Wood,Old Burnhamthorpe,Bloordale Gard...
96       M9L    North York                                      Humber Summit
97       M9M    North York                                    Emery,Humberlea
98       M9N          York                                             Weston
99       M9P     Etobicoke                                          Westmount
100      M9R     Etobicoke  Richview Gardens,Kingsview Village,St. Phillip...
101      M9V     Etobicoke  Mount Olive,South Steeles,Thistletown,Silverst...
102      M9W     Etobicoke                                          Northwest

Now, as the assignment asked, if a cell has a borough but a Not assigned neighborhood, then the neighborhood will be given the same as the borough.

In [8]:
table_temp2.loc[table_temp2['Neighbourhood']=='Not assigned','Neighbourhood']=table_temp2['Borough']

In [9]:
table_temp2.tail(20)

Postcode       Borough                                      Neighbourhood
83       M6R  West Toronto                              Parkdale,Roncesvalles
84       M6S  West Toronto                                  Swansea,Runnymede
85       M7A  Queen's Park                                       Queen's Park
86       M7R   Mississauga              Canada Post Gateway Processing Centre
87       M7Y  East Toronto  Business Reply Mail Processing Centre 969 Eastern
88       M8V     Etobicoke         Mimico South,New Toronto,Humber Bay Shores
89       M8W     Etobicoke                              Alderwood,Long Branch
90       M8X     Etobicoke        The Kingsway,Montgomery Road,Old Mill North
91       M8Y     Etobicoke  Sunnylea,Kingsway Park South East,Mimico NE,Th...
92       M8Z     Etobicoke  Kingsway Park South West,The Queensway West,Mi...
93       M9A     Etobicoke                                   Islington Avenue
94       M9B     Etobicoke  Princess Gardens,West Deane Park,Martin Grove,...
95       M9C     Etobicoke  Markland Wood,Old Burnhamthorpe,Bloordale Gard...
96       M9L    North York                                      Humber Summit
97       M9M    North York                                    Emery,Humberlea
98       M9N          York                                             Weston
99       M9P     Etobicoke                                          Westmount
100      M9R     Etobicoke  Richview Gardens,Kingsview Village,St. Phillip...
101      M9V     Etobicoke  Mount Olive,South Steeles,Thistletown,Silverst...
102      M9W     Etobicoke                                          Northwest

Finally, let's check the shape of the dataframe as required. **Rename the columns** as the assignment used for consistency.

In [10]:
table_cy=table_temp2
table_cy.rename(columns={'Postcode':'PostalCode','Neighbourhood':'Neighborhood'},inplace=True)
table_cy.head()

PostalCode      Borough                          Neighborhood
0        M1B  Scarborough                         Malvern,Rouge
1        M1C  Scarborough  Highland Creek,Rouge Hill,Port Union
2        M1E  Scarborough       Morningside,Guildwood,West Hill
3        M1G  Scarborough                                Woburn
4        M1H  Scarborough                             Cedarbrae

In [11]:
table_cy.shape

(103, 3)

So now it has 103 rows.

### **Part II** -- Get Latitude/Longitude Coordinates

In [12]:
csv='https://cocl.us/Geospatial_data'
Geospatial=pd.read_csv(csv)
Geospatial.rename(columns={'Postal Code':'PostalCode'},inplace=True)
Geospatial

PostalCode   Latitude  Longitude
0          M1B  43.806686 -79.194353
1          M1C  43.784535 -79.160497
2          M1E  43.763573 -79.188711
3          M1G  43.770992 -79.216917
4          M1H  43.773136 -79.239476
..         ...        ...        ...
98         M9N  43.706876 -79.518188
99         M9P  43.696319 -79.532242
100        M9R  43.688905 -79.554724
101        M9V  43.739416 -79.588437
102        M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [13]:
table_use=pd.merge(table_cy,Geospatial,left_on='PostalCode',right_on='PostalCode')
table_use

PostalCode      Borough  \
0          M1B  Scarborough   
1          M1C  Scarborough   
2          M1E  Scarborough   
3          M1G  Scarborough   
4          M1H  Scarborough   
..         ...          ...   
98         M9N         York   
99         M9P    Etobicoke   
100        M9R    Etobicoke   
101        M9V    Etobicoke   
102        M9W    Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                        Malvern,Rouge  43.806686 -79.194353  
1                 Highland Creek,Rouge Hill,Port Union  43.784535 -79.160497  
2                      Morningside,Guildwood,West Hill  43.763573 -79.188711  
3                                               Woburn  43.770992 -79.216917  
4                                            Cedarbrae  43.773136 -79.239476  
..                                                 ...        ...        ...  
98                                              Weston  43.706876 -79.518188  
99                                           Westmount  43.696319 -79.532242  
100  Richview Gardens,Kingsview Village,St. Phillip...  43.688905 -79.554724  
101  Mount Olive,South Steeles,Thistletown,Silverst...  43.739416 -79.588437  
102                                          Northwest  43.706748 -79.594054  

[103 rows x 5 columns]

In [14]:
table_use.shape

(103, 5)

In [15]:
table_use.describe(include='all')

PostalCode     Borough      Neighborhood    Latitude   Longitude
count         103         103               103  103.000000  103.000000
unique        103          11               103         NaN         NaN
top           M4G  North York  Thorncliffe Park         NaN         NaN
freq            1          24                 1         NaN         NaN
mean          NaN         NaN               NaN   43.704608  -79.397153
std           NaN         NaN               NaN    0.052463    0.097146
min           NaN         NaN               NaN   43.602414  -79.615819
25%           NaN         NaN               NaN   43.660567  -79.464763
50%           NaN         NaN               NaN   43.696948  -79.388790
75%           NaN         NaN               NaN   43.745320  -79.340923
max           NaN         NaN               NaN   43.836125  -79.160497

### **Part III** -- Explore and cluster the neighborhoods in Toronto

Before we start exploring, let's download key libraries that we may need to use.

In [16]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library


Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.5.11
  latest version: 4.7.12

Please update conda by running

    $ conda update -n base -c defaults conda



## Package Plan ##

  environment location: /home/jupyterlab/conda/envs/python

  added / updated specs: 
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    geopy-1.20.0               |             py_0          57 KB  conda-forge
    geographiclib-1.49         |             py_0          32 KB  conda-forge
    ------------------------------------------------------------
                                           Total:          90 KB

The following NEW packages will be INSTALLED:

    geographiclib: 1.49-py_0   conda-forge
    geopy:         1.20.0-py_0 conda-forge


geopy-1.20.0         | 57 KB     | ##################################### | 100% 
geographiclib-1.49   | 

In [17]:
table_use.Borough.value_counts()

North York          24
Downtown Toronto    18
Scarborough         17
Etobicoke           12
Central Toronto      9
West Toronto         6
York                 5
East York            5
East Toronto         5
Mississauga          1
Queen's Park         1
Name: Borough, dtype: int64

#### **As asked, now let's select those boroughs that contain the word Toronto.** 

In [18]:
table_tor=table_use[table_use.Borough.str.contains('Toronto')].reset_index(drop=True)
table_tor.head()

PostalCode          Borough                   Neighborhood   Latitude  \
0        M4E     East Toronto                    The Beaches  43.676357   
1        M4K     East Toronto    Riverdale,The Danforth West  43.679557   
2        M4L     East Toronto  The Beaches West,India Bazaar  43.668999   
3        M4M     East Toronto                Studio District  43.659526   
4        M4N  Central Toronto                  Lawrence Park  43.728020   

   Longitude  
0 -79.293031  
1 -79.352188  
2 -79.315572  
3 -79.340923  
4 -79.388790

In [19]:
table_tor

PostalCode           Borough  \
0         M4E      East Toronto   
1         M4K      East Toronto   
2         M4L      East Toronto   
3         M4M      East Toronto   
4         M4N   Central Toronto   
5         M4P   Central Toronto   
6         M4R   Central Toronto   
7         M4S   Central Toronto   
8         M4T   Central Toronto   
9         M4V   Central Toronto   
10        M4W  Downtown Toronto   
11        M4X  Downtown Toronto   
12        M4Y  Downtown Toronto   
13        M5A  Downtown Toronto   
14        M5B  Downtown Toronto   
15        M5C  Downtown Toronto   
16        M5E  Downtown Toronto   
17        M5G  Downtown Toronto   
18        M5H  Downtown Toronto   
19        M5J  Downtown Toronto   
20        M5K  Downtown Toronto   
21        M5L  Downtown Toronto   
22        M5N   Central Toronto   
23        M5P   Central Toronto   
24        M5R   Central Toronto   
25        M5S  Downtown Toronto   
26        M5T  Downtown Toronto   
27        M5V  Downtown Toronto   
28        M5W  Downtown Toronto   
29        M5X  Downtown Toronto   
30        M6G  Downtown Toronto   
31        M6H      West Toronto   
32        M6J      West Toronto   
33        M6K      West Toronto   
34        M6P      West Toronto   
35        M6R      West Toronto   
36        M6S      West Toronto   
37        M7Y      East Toronto   

                                         Neighborhood   Latitude  Longitude  
0                                         The Beaches  43.676357 -79.293031  
1                         Riverdale,The Danforth West  43.679557 -79.352188  
2                       The Beaches West,India Bazaar  43.668999 -79.315572  
3                                     Studio District  43.659526 -79.340923  
4                                       Lawrence Park  43.728020 -79.388790  
5                                    Davisville North  43.712751 -79.390197  
6                                  North Toronto West  43.715383 -79.405678  
7                                          Davisville  43.704324 -79.388790  
8                          Summerhill East,Moore Park  43.689574 -79.383160  
9   Summerhill West,Forest Hill SE,Deer Park,South...  43.686412 -79.400049  
10                                           Rosedale  43.679563 -79.377529  
11                         St. James Town,Cabbagetown  43.667967 -79.367675  
12                               Church and Wellesley  43.665860 -79.383160  
13                           Harbourfront,Regent Park  43.654260 -79.360636  
14                            Ryerson,Garden District  43.657162 -79.378937  
15                                     St. James Town  43.651494 -79.375418  
16                                        Berczy Park  43.644771 -79.373306  
17                                 Central Bay Street  43.657952 -79.387383  
18                             Adelaide,Richmond,King  43.650571 -79.384568  
19    Harbourfront East,Toronto Islands,Union Station  43.640816 -79.381752  
20            Design Exchange,Toronto Dominion Centre  43.647177 -79.381576  
21                      Commerce Court,Victoria Hotel  43.648198 -79.379817  
22                                           Roselawn  43.711695 -79.416936  
23                 Forest Hill West,Forest Hill North  43.696948 -79.411307  
24                  Yorkville,The Annex,North Midtown  43.672710 -79.405678  
25                      University of Toronto,Harbord  43.662696 -79.400049  
26            Kensington Market,Grange Park,Chinatown  43.653206 -79.400049  
27  South Niagara,Bathurst Quay,King and Spadina,R...  43.628947 -79.394420  
28                    Stn A PO Boxes 25 The Esplanade  43.646435 -79.374846  
29              Underground city,First Canadian Place  43.648429 -79.382280  
30                                           Christie  43.669542 -79.422564  
31                        Dovercourt Village,Dufferin  43.669005 -79.442259  
32                            Little Portugal,Trinity  43.64792

In [65]:
table_tor.shape

(38, 5)

In [20]:
address = 'Toronto, Canada'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


In [66]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=12)

# add markers to map
for lat, lng, neigh, borough in zip(table_tor['Latitude'], table_tor['Longitude'], table_tor['Neighborhood'], table_tor['Borough']):
    label = '{}, {}'.format(neigh, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

#### Next Define Foursquare Credentials and Version

In [22]:
CLIENT_ID = 'CLEBLHUA2WPRPAEWW5X4U5IPJ23GG2JPUZT34N0EKX54HVZQ' # your Foursquare ID
CLIENT_SECRET = 'RCCHJNMELZ5VGBOH4M3HXQMGDF2OPY24JXLS1JZ0M25ACXEL' # your Foursquare Secret
VERSION = '20190916' # Foursquare API version

print('My credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

My credentails:
CLIENT_ID: CLEBLHUA2WPRPAEWW5X4U5IPJ23GG2JPUZT34N0EKX54HVZQ
CLIENT_SECRET:RCCHJNMELZ5VGBOH4M3HXQMGDF2OPY24JXLS1JZ0M25ACXEL


#### **As a test, let's check the top 100 venues that are in a downtown neighborhood within a radius of 500 meters.**

In [23]:
table_tor.loc[20, 'Neighborhood']

'Design Exchange,Toronto Dominion Centre'

In [24]:
neighborhood_latitude = table_tor.loc[20, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = table_tor.loc[20, 'Longitude'] # neighborhood longitude value

neighborhood_name = table_tor.loc[20, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Design Exchange,Toronto Dominion Centre are 43.6471768, -79.38157640000001.


In [25]:
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url # display URL


'https://api.foursquare.com/v2/venues/explore?&client_id=CLEBLHUA2WPRPAEWW5X4U5IPJ23GG2JPUZT34N0EKX54HVZQ&client_secret=RCCHJNMELZ5VGBOH4M3HXQMGDF2OPY24JXLS1JZ0M25ACXEL&v=20190916&ll=43.6471768,-79.38157640000001&radius=500&limit=100'

In [26]:
results = requests.get(url).json()
#results

From the Foursquare lab in the previous module, we know that all the information is in the ***items*** key. Before we proceed, let's borrow the **get_category_type** function from the Foursquare lab.

In [27]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Now we are ready to clean the json and structure it into a *pandas* dataframe.

Get the ***items*** key.

In [28]:
venues = results['response']['groups'][0]['items']
#venues

In [29]:
nearby_venues = json_normalize(venues) # flatten JSON
nearby_venues.head(2)

referralId  reasons.count  \
0  e-0-4ad4c05df964a52059f620e3-0              0   
1  e-0-501ae947e4b0d11883b910a7-1              0   

                                       reasons.items  \
0  [{'summary': 'This spot is popular', 'type': '...   
1  [{'summary': 'This spot is popular', 'type': '...   

                   venue.id          venue.name venue.location.address  \
0  4ad4c05df964a52059f620e3               Canoe  66 Wellington St West   
1  501ae947e4b0d11883b910a7  Equinox Bay Street             199 Bay St   

           venue.location.crossStreet  venue.location.lat  venue.location.lng  \
0                       at Bay Street           43.647452          -79.381320   
1  at Commerce Court West, PATH Level           43.648100          -79.379989   

                       venue.location.labeledLatLngs  venue.location.distance  \
0  [{'label': 'display', 'lat': 43.64745206618347...                       36   
1  [{'label': 'display', 'lat': 43.64809974034856...                      164   

  venue.location.postalCode venue.location.cc venue.location.city  \
0                   M5K 1H6                CA             Toronto   
1                   M5L 1L5                CA             Toronto   

  venue.location.state venue.location.country  \
0                   ON                 Canada   
1                   ON                 Canada   

                     venue.location.formattedAddress  \
0  [66 Wellington St West (at Bay Street), Toront...   
1  [199 Bay St (at Commerce Court West, PATH Leve...   

                                    venue.categories  venue.photos.count  \
0  [{'id': '4bf58dd8d48988d1c4941735', 'name': 'R...                   0   
1  [{'id': '4bf58dd8d48988d176941735', 'name': 'G...                   0   

  venue.photos.groups venue.venuePage.id venue.location.neighborhood  
0                  []           40552945                         NaN  
1                  []                NaN                         NaN

In [30]:
# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]
nearby_venues.head(2)

venue.name                                   venue.categories  \
0               Canoe  [{'id': '4bf58dd8d48988d1c4941735', 'name': 'R...   
1  Equinox Bay Street  [{'id': '4bf58dd8d48988d176941735', 'name': 'G...   

   venue.location.lat  venue.location.lng  
0           43.647452          -79.381320  
1           43.648100          -79.379989

In [31]:
# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)
nearby_venues.head(2)

venue.name venue.categories  venue.location.lat  venue.location.lng
0               Canoe       Restaurant           43.647452          -79.381320
1  Equinox Bay Street              Gym           43.648100          -79.379989

In [32]:
# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]
nearby_venues.head(2)

name  categories        lat        lng
0               Canoe  Restaurant  43.647452 -79.381320
1  Equinox Bay Street         Gym  43.648100 -79.379989

In [33]:
nearby_venues.shape

(100, 4)

#### **Now let's repeat the venue checking process for all the neighborhoods.**

In [34]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [35]:
toronto_venues=getNearbyVenues(names=table_tor['Neighborhood'],
                                   latitudes=table_tor['Latitude'],
                                   longitudes=table_tor['Longitude']
                                  )

The Beaches
Riverdale,The Danforth West
The Beaches West,India Bazaar
Studio District
Lawrence Park
Davisville North
North Toronto West
Davisville
Summerhill East,Moore Park
Summerhill West,Forest Hill SE,Deer Park,South Hill,Rathnelly
Rosedale
St. James Town,Cabbagetown
Church and Wellesley
Harbourfront,Regent Park
Ryerson,Garden District
St. James Town
Berczy Park
Central Bay Street
Adelaide,Richmond,King
Harbourfront East,Toronto Islands,Union Station
Design Exchange,Toronto Dominion Centre
Commerce Court,Victoria Hotel
Roselawn
Forest Hill West,Forest Hill North
Yorkville,The Annex,North Midtown
University of Toronto,Harbord
Kensington Market,Grange Park,Chinatown
South Niagara,Bathurst Quay,King and Spadina,Railway Lands,CN Tower,Harbourfront West,Island airport
Stn A PO Boxes 25 The Esplanade
Underground city,First Canadian Place
Christie
Dovercourt Village,Dufferin
Little Portugal,Trinity
Exhibition Place,Parkdale Village,Brockton
High Park,The Junction South
Parkdale,Roncesvall

In [52]:
print(toronto_venues.shape)
toronto_venues.iloc[0:1702:200,:]

(1702, 7)


Neighborhood  Neighborhood Latitude  \
0                                    The Beaches              43.676357   
200                   St. James Town,Cabbagetown              43.667967   
400                      Ryerson,Garden District              43.657162   
600                                  Berczy Park              43.644771   
800                       Adelaide,Richmond,King              43.650571   
1000     Design Exchange,Toronto Dominion Centre              43.647177   
1200     Kensington Market,Grange Park,Chinatown              43.653206   
1400       Underground city,First Canadian Place              43.648429   
1600  Exhibition Place,Parkdale Village,Brockton              43.636847   

      Neighborhood Longitude                     Venue  Venue Latitude  \
0                 -79.293031         Glen Manor Ravine       43.676821   
200               -79.367675                Murgatroid       43.667381   
400               -79.378937             The 3 Brewers       43.655474   
600               -79.373306   FARMR Eatery & Catering       43.648420   
800               -79.384568  Astarté Fresh Yogurt Bar       43.647596   
1000              -79.381576       Toronto PATH System       43.649903   
1200              -79.400049              Golden Patty       43.654659   
1400              -79.382280        Equinox Bay Street       43.648100   
1600              -79.428191                   Caffino       43.639021   

      Venue Longitude        Venue Category  
0          -79.293942                 Trail  
200        -79.369311            Restaurant  
400        -79.380413              Beer Bar  
600        -79.370274                Bistro  
800        -79.386419                  Café  
1000       -79.383053        General Travel  
1200       -79.401179  Caribbean Restaurant  
1400       -79.379989                   Gym  
1600       -79.425289    Italian Restaurant

In [53]:
toronto_venues.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0                  The Beaches              43.676357              -79.293031   
1                  The Beaches              43.676357              -79.293031   
2                  The Beaches              43.676357              -79.293031   
3                  The Beaches              43.676357              -79.293031   
4  Riverdale,The Danforth West              43.679557              -79.352188   

                                Venue  Venue Latitude  Venue Longitude  \
0                   Glen Manor Ravine       43.676821       -79.293942   
1  The Big Carrot Natural Food Market       43.678879       -79.297734   
2                 Grover Pub and Grub       43.679181       -79.297215   
3                       Upper Beaches       43.680563       -79.292869   
4                            Pantheon       43.677621       -79.351434   

      Venue Category  
0              Trail  
1  Health Food Store  
2                Pub  
3       Neighborhood  
4   Greek Restaurant

In [37]:
len(toronto_venues['Venue Category'].unique())

239

In [45]:
#toronto_venues['Venue Category'].value_counts()

In [46]:
#toronto_venues['Venue Category'].unique()

In [54]:
toronto_venues[(toronto_venues['Venue Category'] == 'Neighborhood')]

Neighborhood  Neighborhood Latitude  \
3                                        The Beaches              43.676357   
72                                   Studio District              43.659526   
732                           Adelaide,Richmond,King              43.650571   
820  Harbourfront East,Toronto Islands,Union Station              43.640816   

     Neighborhood Longitude             Venue  Venue Latitude  \
3                -79.293031     Upper Beaches       43.680563   
72               -79.340923       Leslieville       43.662070   
732              -79.384568  Downtown Toronto       43.653232   
820              -79.381752      Harbourfront       43.639526   

     Venue Longitude Venue Category  
3         -79.292869   Neighborhood  
72        -79.337856   Neighborhood  
732       -79.385296   Neighborhood  
820       -79.380688   Neighborhood

In [55]:
toronto_venues.loc[toronto_venues['Venue Category']=='Neighborhood']

Neighborhood  Neighborhood Latitude  \
3                                        The Beaches              43.676357   
72                                   Studio District              43.659526   
732                           Adelaide,Richmond,King              43.650571   
820  Harbourfront East,Toronto Islands,Union Station              43.640816   

     Neighborhood Longitude             Venue  Venue Latitude  \
3                -79.293031     Upper Beaches       43.680563   
72               -79.340923       Leslieville       43.662070   
732              -79.384568  Downtown Toronto       43.653232   
820              -79.381752      Harbourfront       43.639526   

     Venue Longitude Venue Category  
3         -79.292869   Neighborhood  
72        -79.337856   Neighborhood  
732       -79.385296   Neighborhood  
820       -79.380688   Neighborhood

#### **Next let's analyze each neighborhood.**

In [56]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")
toronto_onehot.head()

Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                  0        0                   0             0   
1                  0        0                   0             0   
2                  0        0                   0             0   
3                  0        0                   0             0   
4                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0         0            0           0                    0   
1             0         0            0           0                    0   
2             0         0            0           0                    0   
3             0         0            0           0                    0   
4             0         0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  \
0                 0                   0              0          0           0   
1                 0                   0              0          0           0   
2                 0                   0              0          0           0   
3                 0                   0              0          0           0   
4                 0                   0              0          0           0   

   Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  \
0           0       0     0    0                 0                   0      0   
1           0       0     0    0                 0                   0      0   
2           0       0     0    0                 0                   0      0   
3           0       0     0    0                 0                   0      0   
4           0       0     0    0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bistro  \
0                0         0           0                   0       0   
1                0         0           0                   0       0   
2                0         0           0                   0       0   
3                0         0           0                   0       0   
4                0         0           0                   0       0   

   Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  \
0              0          0         0                     0               0   
1              0          0         0                     0               0   
2              0          0         0                     0               0   
3              0          0         0                     0               0   
4              0          0         0                     0               0   

   Brewery  Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \
0        0                0         0             0              0         0   
1        0                0         0             0              0         0   
2        0                0         0             0              0         0   
3        0                0         0             0              0         0   
4        0                0         0             0              0         0   

   Butcher  Café  Cajun / Creole Restaurant  Camera Store  \
0        0     0                          0             0   
1        0     0                          0             0   
2        0     0                          0             0   
3        0     0                          0             0   
4        0     0  

In [57]:
# add neighborhood column back to dataframe
toronto_onehot['TorNeighborhood'] = toronto_venues['Neighborhood'] 
toronto_onehot.head()

Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                  0        0                   0             0   
1                  0        0                   0             0   
2                  0        0                   0             0   
3                  0        0                   0             0   
4                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0         0            0           0                    0   
1             0         0            0           0                    0   
2             0         0            0           0                    0   
3             0         0            0           0                    0   
4             0         0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  \
0                 0                   0              0          0           0   
1                 0                   0              0          0           0   
2                 0                   0              0          0           0   
3                 0                   0              0          0           0   
4                 0                   0              0          0           0   

   Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  \
0           0       0     0    0                 0                   0      0   
1           0       0     0    0                 0                   0      0   
2           0       0     0    0                 0                   0      0   
3           0       0     0    0                 0                   0      0   
4           0       0     0    0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bistro  \
0                0         0           0                   0       0   
1                0         0           0                   0       0   
2                0         0           0                   0       0   
3                0         0           0                   0       0   
4                0         0           0                   0       0   

   Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  \
0              0          0         0                     0               0   
1              0          0         0                     0               0   
2              0          0         0                     0               0   
3              0          0         0                     0               0   
4              0          0         0                     0               0   

   Brewery  Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \
0        0                0         0             0              0         0   
1        0                0         0             0              0         0   
2        0                0         0             0              0         0   
3        0                0         0             0              0         0   
4        0                0         0             0              0         0   

   Butcher  Café  Cajun / Creole Restaurant  Camera Store  \
0        0     0                          0             0   
1        0     0                          0             0   
2        0     0                          0             0   
3        0     0                          0             0   
4        0     0  

In [58]:
# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]
toronto_onehot.rename(columns={'TorNeighborhood':'CityNeighborhood'},inplace=True)

In [59]:
print(toronto_onehot.shape)
toronto_onehot.loc[0:1700:200,:]

(1702, 240)


CityNeighborhood  Afghan Restaurant  Airport  \
0                                    The Beaches                  0        0   
200                   St. James Town,Cabbagetown                  0        0   
400                      Ryerson,Garden District                  0        0   
600                                  Berczy Park                  0        0   
800                       Adelaide,Richmond,King                  0        0   
1000     Design Exchange,Toronto Dominion Centre                  0        0   
1200     Kensington Market,Grange Park,Chinatown                  0        0   
1400       Underground city,First Canadian Place                  0        0   
1600  Exhibition Place,Parkdale Village,Brockton                  0        0   

      Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                      0             0               0                0   
200                    0             0               0                0   
400                    0             0               0                0   
600                    0             0               0                0   
800                    0             0               0                0   
1000                   0             0               0                0   
1200                   0             0               0                0   
1400                   0             0               0                0   
1600                   0             0               0                0   

      Airport Terminal  American Restaurant  Antique Shop  Aquarium  \
0                    0                    0             0         0   
200                  0                    0             0         0   
400                  0                    0             0         0   
600                  0                    0             0         0   
800                  0                    0             0         0   
1000                 0                    0             0         0   
1200                 0                    0             0         0   
1400                 0                    0             0         0   
1600                 0                    0             0         0   

      Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
0               0           0                    0                 0   
200             0           0                    0                 0   
400             0           0                    0                 0   
600             0           0                    0                 0   
800             0           0                    0                 0   
1000            0           0                    0                 0   
1200            0           0                    0                 0   
1400            0           0                    0                 0   
1600            0           0                    0                 0   

      Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  \
0                      0              0          0           0           0   
200                    0              0          0           0           0   
400                    0              0          0           0           0   
600                    0              0          0           0           0   
800                    0              0          0           0           0   
1000                   0              0          0           0           0   
1200                   0              0          0           0           0   
1400                   0              0          0           0           0   
1600                   0              0          0           0           0   

      Bakery  Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  \
0          0     0    0                 0                   0      0   
200        0     0    0                 0                   0      0   
400        0     0    0                 0          

#### **Now let's group rows by neighborhood and by taking the means of the frequency of occurrence of each category.**

In [69]:
toronto_onehot.head()
toronto_group = toronto_onehot.groupby('CityNeighborhood').mean().reset_index()
toronto_group.loc[0:38:5,:]

CityNeighborhood  Afghan Restaurant  \
0                              Adelaide,Richmond,King           0.000000   
5                                Church and Wellesley           0.011111   
10                        Dovercourt Village,Dufferin           0.000000   
15                       High Park,The Junction South           0.000000   
20                              Parkdale,Roncesvalles           0.000000   
25  South Niagara,Bathurst Quay,King and Spadina,R...           0.000000   
30                         Summerhill East,Moore Park           0.000000   
35              Underground city,First Canadian Place           0.000000   

    Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0    0.0000              0.0000        0.0000           0.000   
5    0.0000              0.0000        0.0000           0.000   
10   0.0000              0.0000        0.0000           0.000   
15   0.0000              0.0000        0.0000           0.000   
20   0.0000              0.0000        0.0000           0.000   
25   0.0625              0.0625        0.0625           0.125   
30   0.0000              0.0000        0.0000           0.000   
35   0.0000              0.0000        0.0000           0.000   

    Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0             0.000             0.000             0.030000           0.0   
5             0.000             0.000             0.011111           0.0   
10            0.000             0.000             0.000000           0.0   
15            0.000             0.000             0.000000           0.0   
20            0.000             0.000             0.000000           0.0   
25            0.125             0.125             0.000000           0.0   
30            0.000             0.000             0.000000           0.0   
35            0.000             0.000             0.030000           0.0   

    Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
0        0.0         0.01        0.01             0.000000              0.02   
5        0.0         0.00        0.00             0.011111              0.00   
10       0.0         0.00        0.00             0.000000              0.00   
15       0.0         0.00        0.00             0.045455              0.00   
20       0.0         0.00        0.00             0.000000              0.00   
25       0.0         0.00        0.00             0.000000              0.00   
30       0.0         0.00        0.00             0.000000              0.00   
35       0.0         0.01        0.00             0.000000              0.03   

    Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  \
0                  0.0            0.0        0.0         0.0         0.0   
5                  0.0            0.0        0.0         0.0         0.0   
10                 0.0            0.0        0.0         0.0         0.0   
15                 0.0            0.0        0.0         0.0         0.0   
20                 0.0            0.0        0.0         0.0         0.0   
25                 0.0            0.0        0.0         0.0         0.0   
30                 0.0            0.0        0.0         0.0         0.0   
35                 0.0            0.0        0.0         0.0         0.0   

      Bakery      Bank       Bar  Baseball Stadium  Basketball Stadium  Beach  \
0   0.020000  0.000000  0.040000               0.0                 0.0    0.0   
5   0.000000  0.000000  0.000000               0.0                 0.0    0.0   
10  0.125000  0.062500  0.062500               0.0                 0.0    0.0   
15  0.045455  0.000000  0.045455               0.0                 0.0    0.0   
20  0.000000  0.066667  0.066667               0.0                 0.0    0.0   
25  0.000000  0.000000  0.062500               0.0                 0.0    0.0   
30  0.000000  0.000000  0.000000               0.0                 0.0    0.0   
35  0.020000  0.000000  0.0

In [63]:
toronto_group.shape

(38, 240)

#### Let's print each neighborhood along with the top 5 most common venues

In [92]:
num_top_venues = 5
count=0

for hood in toronto_group['CityNeighborhood']:
    print("----"+hood+"----")
    temp = toronto_group[toronto_group['CityNeighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    count=count+1
    if (count-1)%10==0:
        print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
        print('CityNeighborhood sequence number as ',count-1,'\n')
    else:
        print('Note here only print every 10th result to check',' \n')

----Adelaide,Richmond,King----
         venue  freq
0  Coffee Shop  0.08
1         Café  0.05
2          Bar  0.04
3   Steakhouse  0.04
4   Restaurant  0.03
CityNeighborhood sequence number as  0 

----Berczy Park----
Note here only print every 10th result to check  

----Business Reply Mail Processing Centre 969 Eastern----
Note here only print every 10th result to check  

----Central Bay Street----
Note here only print every 10th result to check  

----Christie----
Note here only print every 10th result to check  

----Church and Wellesley----
Note here only print every 10th result to check  

----Commerce Court,Victoria Hotel----
Note here only print every 10th result to check  

----Davisville----
Note here only print every 10th result to check  

----Davisville North----
Note here only print every 10th result to check  

----Design Exchange,Toronto Dominion Centre----
Note here only print every 10th result to check  

----Dovercourt Village,Dufferin----
                  venue  f

#### Let's put that into a *pandas* dataframe

First, let's write a function to sort the venues in descending order.

In [93]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [100]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['CityNeighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['CityNeighborhood'] = toronto_group['CityNeighborhood']

for ind in np.arange(toronto_group.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_group.iloc[ind, :], num_top_venues)

print('Dataframe shape is',neighborhoods_venues_sorted.shape)
neighborhoods_venues_sorted.loc[0:38:5,:]

Dataframe shape is (38, 11)


CityNeighborhood 1st Most Common Venue  \
0                              Adelaide,Richmond,King           Coffee Shop   
5                                Church and Wellesley           Coffee Shop   
10                        Dovercourt Village,Dufferin                Bakery   
15                       High Park,The Junction South                  Café   
20                              Parkdale,Roncesvalles        Breakfast Spot   
25  South Niagara,Bathurst Quay,King and Spadina,R...        Airport Lounge   
30                         Summerhill East,Moore Park                  Park   
35              Underground city,First Canadian Place           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue   4th Most Common Venue  \
0                   Café            Steakhouse                     Bar   
5                Gay Bar   Japanese Restaurant        Sushi Restaurant   
10              Pharmacy           Supermarket             Pizza Place   
15    Mexican Restaurant           Flea Market  Furniture / Home Store   
20             Gift Shop             Bookstore                    Bank   
25       Airport Service      Airport Terminal                   Plane   
30          Tennis Court           Yoga Studio          Discount Store   
35                  Café                 Hotel              Restaurant   

   5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
0           Burger Joint            Restaurant               Cosmetics Shop   
5             Restaurant                   Gym                 Dance Studio   
10           Coffee Shop           Music Venue                         Café   
15                 Diner           Music Venue               Discount Store   
20         Movie Theater            Restaurant  Eastern European Restaurant   
25      Sculpture Garden           Coffee Shop                Boat or Ferry   
30              Festival  Fast Food Restaurant               Farmers Market   
35            Steakhouse             Gastropub          American Restaurant   

        8th Most Common Venue 9th Most Common Venue    10th Most Common Venue  
0             Thai Restaurant   American Restaurant                     Hotel  
5             Bubble Tea Shop             Gastropub  Mediterranean Restaurant  
10  Middle Eastern Restaurant               Brewery                       Bar  
15  Cajun / Creole Restaurant  Fast Food Restaurant                 Bookstore  
20         Italian Restaurant               Dog Run          Cuban Restaurant  
25            Harbor / Marina              Boutique              Airport Gate  
30         Falafel Restaurant           Event Space      Ethiopian Restaurant  
35         Seafood Restaurant                   Bar              Burger Joint

## Now run K-Means Clustering to cluster the neighborhood into 8 clusters.

In [126]:
# set number of clusters
kclusters = 10

toronto_group_clustering = toronto_group.drop('CityNeighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_group_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 8, 0], dtype=int32)

In [127]:
neighborhoods_venues_sorted.drop(columns='Cluster Labels',inplace=True)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [128]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = table_tor

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('CityNeighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

PostalCode          Borough                   Neighborhood   Latitude  \
0        M4E     East Toronto                    The Beaches  43.676357   
1        M4K     East Toronto    Riverdale,The Danforth West  43.679557   
2        M4L     East Toronto  The Beaches West,India Bazaar  43.668999   
3        M4M     East Toronto                Studio District  43.659526   
4        M4N  Central Toronto                  Lawrence Park  43.728020   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.293031               4     Health Food Store                 Trail   
1 -79.352188               0      Greek Restaurant           Coffee Shop   
2 -79.315572               0  Fast Food Restaurant        Ice Cream Shop   
3 -79.340923               0                  Café           Coffee Shop   
4 -79.388790               2                  Park              Bus Line   

  3rd Most Common Venue   4th Most Common Venue 5th Most Common Venue  \
0                   Pub            Neighborhood        Farmers Market   
1    Italian Restaurant  Furniture / Home Store        Ice Cream Shop   
2          Burger Joint            Liquor Store        Sandwich Place   
3                Bakery      Italian Restaurant   American Restaurant   
4           Swim School             Yoga Studio               Dog Run   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0    Falafel Restaurant           Event Space  Fast Food Restaurant   
1             Bookstore           Yoga Studio               Brewery   
2         Burrito Place     Fish & Chips Shop            Steakhouse   
3    Chinese Restaurant                   Bar      Stationery Store   
4              Festival  Fast Food Restaurant        Farmers Market   

  9th Most Common Venue     10th Most Common Venue  
0        Discount Store          Electronics Store  
1       Bubble Tea Shop                 Restaurant  
2    Italian Restaurant           Sushi Restaurant  
3        Clothing Store  Middle Eastern Restaurant  
4    Falafel Restaurant                Event Space

#### Finally let's visualize the resulting clusters.

In [131]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=12)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=10,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-2],
        fill_opacity=2).add_to(map_clusters)
       
map_clusters
#map_toronto

## Compare with the original Toronto neighborhood map as a double check.

In [108]:
map_toronto

# ***Thank You!***